### PR TITLE
v2 improvements

### PR DESCRIPTION
**Pull request by snapshotpl**
* Make all implementation final,
* add typed args where possible,
* add suffix handler named constructor,
* support for php 7.4

### CODE REVIEW

File: src/Handler/CanNotInvokeHandler.php [hunk 1]
```diff
@@ -39,12 +35,15 @@ public static function forCommand($command, string $reason)
         return $exception;
     }
 
+    private function __construct(string $message = '', $code = 0, Throwable $previous = null)
+    {
+        parent::__construct($message, $code, $previous);
+    }
+
```
Comment (rosstuck) [inline src/Handler/CanNotInvokeHandler.php]: I like the intent here but I prefer to leave the constructor open so other folks can potentially use it down stream. Maybe they'd like to format something entirely different here but still use a Tactician level exception. :)

File: src/Handler/CanNotInvokeHandler.php [hunk 2]
```diff
@@ -39,12 +35,15 @@ public static function forCommand($command, string $reason)
...
     /**
      * Returns the command that could not be invoked
-     *
-     * @return mixed
      */
-    public function getCommand()
+    public function getCommand(): object
     {
         return $this->command;
     }
```
Comment (rosstuck) [inline src/Handler/CanNotInvokeHandler.php]: Other than that one thing, I love everything you've done in this file! :heart: :rainbow: :unicorn: :star2:

File: src/Handler/CommandHandlerMiddleware.php [hunk 3]
```diff
@@ -13,7 +13,7 @@
 /**
  * The "core" CommandBus. Locates the appropriate handler and executes command.
  */
-class CommandHandlerMiddleware implements Middleware
+final class CommandHandlerMiddleware implements Middleware
 {
     /** @var ContainerInterface */
```
Comment (rosstuck) [inline src/Handler/CommandHandlerMiddleware.php]: Groovy! I need to reset my IDE config, le sigh.

File: src/Handler/Mapping/ClassName/Suffix.php [hunk 4]
```diff
@@ -4,11 +4,18 @@
...
+    private const DEFAULT = 'Handler';
+
     /** @var string */
     private $suffix;
 
+    public static function handler(): self
+    {
+        return new self(self::DEFAULT);
+    }
+
```
Comment (rosstuck) [inline src/Handler/Mapping/ClassName/Suffix.php]: Dig the final and while I like the idea, I'd like to avoid supporting too many direct defaults in T2, so would prefer to drop this.

File: src/Handler/Mapping/MethodName/Handle.php [hunk 5]
```diff
@@ -7,7 +7,7 @@
 /**
  * Handle command by calling the "handle" method.
  */
-class Handle implements MethodNameInflector
+final class Handle implements MethodNameInflector
 {
     /**
      * {@inheritdoc}
```
Comment (rosstuck) [inline src/Handler/Mapping/MethodName/Handle.php]: Also groovy! I should just setup my own coding standards thing, similar to doctrine-coding-standards

File: src/Handler/Mapping/MethodName/HandleClassNameWithoutSuffix.php [hunk 6]
```diff
@@ -34,11 +36,12 @@ public function __construct(string $suffix = 'Command')
     {
         $this->suffix       = $suffix;
         $this->suffixLength = strlen($suffix);
+        $this->handleLastPartOfClassName = new HandleLastPartOfClassName();
     }
 
     public function getMethodName(string $commandClassName) : string
     {
-        $methodName = parent::getMethodName($commandClassName);
+        $methodName = $this->handleLastPartOfClassName->getMethodName($commandClassName);
```
Comment (rosstuck) [inline src/Handler/Mapping/MethodName/HandleClassNameWithoutSuffix.php]: I normally avoid constructing anything in the constructor but in this case I'll allow it. :ok_woman: :heavy_check_mark:

File: src/Handler/Mapping/MethodName/HandleLastPartOfClassName.php [hunk 7]
```diff
@@ -13,14 +13,22 @@
...
+final class HandleLastPartOfClassName implements MethodNameInflector
 {
+    /** @var LastPartOfClassName */
+    private $lastPartOfClassName;
+
+    public function __construct()
+    {
+        $this->lastPartOfClassName = new LastPartOfClassName();
+    }
+
```
Comment (rosstuck) [inline src/Handler/Mapping/MethodName/HandleLastPartOfClassName.php]: Same as above :)

File: src/Handler/Mapping/MethodName/Invoke.php [hunk 8]
```diff
@@ -8,7 +8,7 @@
  * Handle command by calling the __invoke magic method. Handy for single
  * use classes or closures.
  */
-class Invoke implements MethodNameInflector
+final class Invoke implements MethodNameInflector
```
Comment (rosstuck) [inline src/Handler/Mapping/MethodName/Invoke.php]: Triple groovy! I should buy a boat.

File: src/Handler/Mapping/MethodName/LastPartOfClassName.php [hunk 9]
```diff
@@ -16,7 +16,7 @@
  *  - \MyGlobalCommand    => $handler->myGlobalCommand()
  *  - \My\App\CreateUser  => $handler->createUser()
  */
-class LastPartOfClassName implements MethodNameInflector
+final class LastPartOfClassName implements MethodNameInflector
```
Comment (rosstuck) [inline src/Handler/Mapping/MethodName/LastPartOfClassName.php]: Quad groo-, you know what, you get the idea

File: tests/Handler/Mapping/ClassName/SuffixTest.php [hunk 10]
```diff
@@ -12,12 +12,19 @@ class SuffixTest extends TestCase
...
         );
     }
 
+    public function testBuildHandlerSuffix(): void
+    {
+        $result = Suffix::handler()->getClassName('MyCommand');
+
+        self::assertSame('MyCommandHandler', $result);
+    }
+
```
Comment (rosstuck) [inline tests/Handler/Mapping/ClassName/SuffixTest.php]: In line with previous comment about ::handle this would go too, sorry :crying_cat_face:

File: tests/Handler/Mapping/ClassName/SuffixTest.php [hunk 11]
```diff
@@ -12,12 +12,19 @@ class SuffixTest extends TestCase
     /** @dataProvider examples */
     public function testCanAddSuffixToClassName(string $suffix, string $command, string $expectedResult) : void
     {
-        self::assertEquals(
+        self::assertSame(
```
Comment (rosstuck) [inline tests/Handler/Mapping/ClassName/SuffixTest.php]: No complaints but curious why the change, just general strictness? :smile_cat:

Comment (snapshotpl) [reply tests/Handler/Mapping/ClassName/SuffixTest.php]: Strictness is safely :)

File: src/Handler/CanNotInvokeHandler.php (hunk 1 above)
Comment (snapshotpl) [reply src/Handler/CanNotInvokeHandler.php]: Look at `command` property. And getter. After one constructor instance will be broken.

Comment (rosstuck) [reply src/Handler/CanNotInvokeHandler.php]: Yeah, there is no way to set it short of doing something dumb like a protected setter. Alright, cool, let's keep it :)

File: tests/Handler/Mapping/ClassName/SuffixTest.php (hunk 11 above)
Comment (rosstuck) [reply tests/Handler/Mapping/ClassName/SuffixTest.php]: Or the illusion of safety sometimes, but duly noted :+1:

**Comment by rosstuck** [comment]
Thanks a lot for all your work on this! I really appreciate it :)

**Comment by snapshotpl** [comment]
Thanks for your review! BTW what are you think about make `handle(): void` in `CommandBus`?

**Comment by rosstuck** [comment]
Tactician explicitly allows return types (because synchronous commands aren't a bad thing), so `void` return type is a no-go. For extra safety, I've been working on phpstan-tactician which supports checking this, even if PHP's type system doesn't support it.

**Comment by snapshotpl** [comment]
@rosstuck do you need some help with v2 release?

**Comment by rosstuck** [comment]
Thanks for asking! The Tactician-PHPStan stuff is working so I'm mainly punching through the existing packages and updating things. Tactician-Logger and Tactician-Doctrine are updated, I'll deprecate Bernard and Command-Events. The bundle is the big challenge, I'm investigating options for that now. After that, need to update all the new docs and then tag all the releases. If the bundle work takes too long, I'll just tag the rest without but would prefer to do things in one fell swoop.

Not sure if any of that sounds interesting but feel free to let me know if something does. :)
